### DISCUSSION TITLE
Rule 3.3. Colocate related fields in custom types – is there a solution 3?

### DISCUSSION THREAD

**Discussion by nareshbhatia** [graphql-rules/graphql-rules#122] [Q&A] [answered]
Rule 3.3. provides two solutions:
1. Create a new type with related fields
2. Use union types with fragments

I believe there is a 3rd solution that has aspects of 1 & 2, but does not use `implements`:

```graphql
type Claim {
  text: String!
  source: ClaimSource
}

type ClaimSourcePhone {
  phone: String!
  operatorCode: String!
}

type ClaimSourceMail {
  email: String!
}

union ClaimSource = ClaimSourcePhone | ClaimSourceMail
```

Is this a good solution or there are any gotcha's in doing it this way?

**Comment by RIP21** [comment, marked as answer]
Yeah, this is quite a valid solution too. 

Mostly depends on what UI expects, or how it's stored in DB.

If in DB it's a flat structure (just a row in the DB) with nullable fields and some sort of `type` field that helps hint stuff in the application layer. Then converting that into a GraphQL response will be easier without nesting just implementing `isTypeOf` or a similar mechanism by checking the `type` field.

If DB structure has nothing to do with it, as it's mostly artificial/calculated type or you use Document-based DB where nested structures are the norm, then your option may more preferable. 

But TBH, by adding field resolvers you can remap from a flat structure to a nested structure quite easily too :) 

So, long story short. Yes, it's a 3rd solution indeed :)

**Reply by nareshbhatia** [reply]
Perfect! Thank you for the clarity, @RIP21.

**Comment by RIP21** [comment]
But TBH it's just rule 3.3, solution 2. but applied on the field layer :)
